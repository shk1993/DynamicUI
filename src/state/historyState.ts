import { IQuestionData } from '../interfaces/questions';

/** the history object. stores all answers the user has made on the current event */
let history = {};

/**
 * called everytime the form is rendered. This function retrieves all the values from react-final-form and sets them in the history object
 * @param values all the values in the history form state
 */
export const setHistory = (values: IQuestionData) => {
  history = values;
};

/**
 * Gets the current history value based on the Id received as a parameter
 * @param Id QuestionID in the history object to retrieve
 * @returns  The value saved in the history object at id
 */
export const getHistoryValues = (Id: string) => {
  return history[Id];
};

/**
 * Checks if the value saved in the history is the same as the value passed in
 * @param id    QuestionID in the history object to determine if value passed in is the same
 * @param value The value to check if it equals the value saved in history
 * @returns     True if the value saved in history is the same as the value parameter
 */
export const isEqual = (id: string, value: string) => {
  return history[id] === value || history[id].includes(value);
};
