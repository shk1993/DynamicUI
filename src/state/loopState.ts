import { Stack } from 'immutable';

/** the loop state object. This object is an object of key-stacks where key is the questionID the stack is associated with */
let loopStateObject = {};

/**
 * The push function will push the value passed in to the stack associated with the questionID in the loop state object
 * @param id QuestionID to push into
 * @param value value to push into the stack of the questionID
 */
const push = (id: string, value: string) => {
  loopStateObject[id] = loopStateObject[id].push(value);
};

// TODO: Do we need this function? I added the addStackToLoop function and I'm not sure
// if we need this one or not
/**
 * The addToLoop function will push the value passed in to the stack by calling the push function
 * @param id    QuestionID to push into
 * @param value value to push into the stack of the questionID
 */
export const addToLoop = (id: string, value: string) => {
  if (id in loopStateObject) {
    push(id, value);
  } else {
    const newVal = {};
    newVal[id] = Stack<string>([value]);
    loopStateObject = { ...loopStateObject, ...newVal };
  }
};

/**
 * Add to the loopState object the id-stack object passed in. This will any already assigned id in the loop state object
 * @param id    QuestionID to push into
 * @param stack stack to add to the loop state by assigning the questionID to the passed in stack
 */
export const addStackToLoop = (id: string, stack: Stack<string>) => {
  const newVal = {};
  newVal[id] = Stack<string>(stack);
  loopStateObject = { ...loopStateObject, ...newVal };
};

/**
 * checks the loop state object to determine if the stack associated with the questionID is empty
 * @param id QuestionID to determine if the stack is empty or not
 * @returns  true if either the stack associated with the questionID is empty, or doesn't exist. False otherwise
 */
export const isLoopEmpty = (id: string) => {
  if (!(id in loopStateObject) || loopStateObject[id].size === 0) {
    return true;
  } else {
    return false;
  }
};

/**
 * Pops an element off of the loop state object from the questionID passed in as a parameter
 * @param id QuestionID to pop from
 */
export const popFromLoop = (id: string) => {
  loopStateObject[id] = loopStateObject[id].pop();
};

/**
 * Returns the element at the top of the stack associated with the QuestionID received as a parameter
 * @param id QuestionID to peek from
 * @returns  Element at the top of the stack at id
 */
export const peekFromLoop = (id: string): string => loopStateObject[id].peek();
