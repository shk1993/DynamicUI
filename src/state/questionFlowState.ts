import { Stack } from 'immutable';
import InitialQuestion from '../util/InitialQuestion';

/** the questionFlow variable. represented as a stack of arrays of questionIDs */
let questionFlow = Stack<string[]>([InitialQuestion()]);

/**
 * Function to push onto the questionFlow stack
 * @param question array of questionID strings to push onto the stack
 */
export const flowPush = (question: string[]) => {
  questionFlow = questionFlow.push(question);
};

/**
 * Function to pop off of the questionFlow stack. Does NOT return the popped element
 */
export const flowPop = () => {
  questionFlow = questionFlow.pop();
};

/**
 * Function for viewing the element in the questionFlow stack
 * @returns the last element pushed in the stack.
 */
export const flowPeek = () => {
  return questionFlow.peek();
};

/**
 * Function to retrieve the number of elements in the questionFlow stack
 */
export const count = () => {
  return questionFlow.count();
};

/**
 * Function that'll return true if the questionID exists in the questionFlow stack
 * @param question the questionID to determine if it exists in the stack
 */
export const flowContains = (question: string) =>
  questionFlow
    .toArray()
    .join(',')
    .split(',')
    .indexOf(question) !== -1;

/**
 * Returns the question flow state as an array of arrays in stack reverse order
 * @returns the question flow state as an array of arrays from oldest to most recent
 */
export const flowReverse = (): string[][] => questionFlow.reverse().toArray();
