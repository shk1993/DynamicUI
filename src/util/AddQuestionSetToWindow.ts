import QuestionSet from '../util/SampleQuestionSet';

declare global {
  // tslint:disable-next-line:interface-name
  interface Window {
    QuestionSet: typeof QuestionSet;
    ShowQuestionSet: () => void;
  }
}

/**
 * Appends the entire QuestionSet JSON to document.body
 */
const showQuestionSet = () => {
  const pre = document.createElement('pre');
  pre.setAttribute('id', 'questionset');
  pre.innerHTML = JSON.stringify(QuestionSet, undefined, 2);
  document.body.appendChild(pre);
};

/**
 * Exports the additional components to the global window variable
 */
export default () => {
  window.QuestionSet = window.QuestionSet || QuestionSet;
  window.ShowQuestionSet = window.ShowQuestionSet || showQuestionSet;
};
