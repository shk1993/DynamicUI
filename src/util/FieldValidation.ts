import { IAttribute } from '../interfaces/attributes';
import { IValidator } from '../interfaces/validator';

/**
 * This function returns a function that reduces all validators.
 * @param validators   array of validator functions
 * @returns A function that reduces the result of array of validtors
 */
const composeValidators = (...validators: IValidator[]) => (
  value: string | undefined,
) =>
  validators.reduce((error, validator) => error || validator(value), undefined);

/**
 * This function filters validation key from the attributes.
 * It returns the regex validator function if validation attribute is present or returns undefined
 * @param attributes   Attributes on the render field
 * @returns       A function that validates field value matches the regex value or undefined
 */
const getRegexValidator = (attributes: IAttribute[]) => {
  const validationAttr = attributes.filter(
    (attr: IAttribute) => attr.key === 'validation',
  );
  if (
    validationAttr &&
    validationAttr.length !== 0 &&
    validationAttr[0].value
  ) {
    const regex = new RegExp(validationAttr[0].value);
    return (value: string | undefined) =>
      !value || regex.test(value) ? undefined : 'Invalid Input';
  }
  return undefined;
};

/**
 * This function filters required key from the attributes.
 * It returns the required validator function required attribute value is true or returns undefined
 * @param attributes   Attributes on the render field
 * @returns       A function that validates field value is required or undefined
 */
const getRequiredValidator = (attributes: IAttribute[]) => {
  const reqAttr = attributes.filter(
    (attr: IAttribute) => attr.key === 'required',
  );
  if (
    reqAttr &&
    reqAttr.length !== 0 &&
    reqAttr[0].value &&
    reqAttr[0].value === 'true'
  ) {
    return (value: string | undefined) => (value ? undefined : 'Required');
  }
  return undefined;
};

/**
 * This function gets the regex validator, required validator based on the attributes and returns a reduced combimned function.
 * @param attributes   Attributes on the render field
 * @returns       A function that reduces errors on all validators
 */
const composeAllValidators = (attributes: IAttribute[]) => {
  let validatorFunctions: IValidator[] = [];
  if (attributes) {
    const regexFunc = getRegexValidator(attributes);
    const reqFunc = getRequiredValidator(attributes);
    if (regexFunc) {
      validatorFunctions = validatorFunctions.concat([regexFunc]);
    }
    if (reqFunc) {
      validatorFunctions = validatorFunctions.concat([reqFunc]);
    }
  }
  return composeValidators(...validatorFunctions);
};

/**
 * This function is the function to get the validators on the field. It returns a function that reduces errors on all validators.
 * @param attributes   Attributes on the render field
 * @returns       A function that reduces errors on all validators
 */
export const getValidators = (attributes: IAttribute[] | undefined) => {
  return attributes ? composeAllValidators(attributes) : undefined;
};
