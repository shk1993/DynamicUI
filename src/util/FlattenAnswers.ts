import { IAnswers } from '../interfaces/answers';

/**
 * Retrieves an object of answers, maps over all the keys, and then returns an array of answer objects
 * @param answers An object of key-object pairs, where key is the question id
 * @returns       An array of answer objects
 */
export default (answers: IAnswers) => {
  return Object.keys(answers).map(ans => answers[ans]);
};
