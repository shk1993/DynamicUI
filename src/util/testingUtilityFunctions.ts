export const silenceConsoleErrors = () => {
  /* silence -- Warning: Stateless function components cannot be given refs. Attempts to access this ref will fail. */
  /* tslint:disable-next-line */
  console.error = () => {};
  /* tslint:disable-next-line */
  console.warn = () => {};
};
