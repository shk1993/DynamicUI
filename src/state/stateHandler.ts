import { getHistoryValues, isEqual, setHistory } from './historyState';
import {
  addStackToLoop,
  addToLoop,
  isLoopEmpty,
  peekFromLoop,
  popFromLoop,
} from './loopState';
import {
  count,
  flowContains,
  flowPeek,
  flowPop,
  flowPush,
  flowReverse,
} from './questionFlowState';

/**
 * Map of the state functions from their state implementations to this shared interface to dynamically call
 */
export const map = {
  flow: {
    contains: flowContains,
    peek: flowPeek,
    pop: flowPop,
    push: flowPush,
    count,
    reverseArray: flowReverse,
  },
  loop: {
    isEmpty: isLoopEmpty,
    peek: peekFromLoop,
    pop: popFromLoop,
    push: addStackToLoop,
    pushOne: addToLoop,
  },
  // tslint:disable-next-line:object-literal-key-quotes
  history: { '==': isEqual, get: getHistoryValues, set: setHistory },
};
