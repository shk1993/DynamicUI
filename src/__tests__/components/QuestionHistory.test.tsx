import * as React from 'react';
import { cleanup, render } from 'react-testing-library';
import {
  changeQuestion,
  QuestionHistory,
} from '../../components/QuestionHistory';
import { flowPeek, flowPush } from '../../state/questionFlowState';
import { silenceConsoleErrors } from '../../util/testingUtilityFunctions';

beforeAll(() => silenceConsoleErrors());
afterEach(cleanup);
const changeFlow = jest.fn();

describe('QuestionHistory component', () => {
  it('checks whether the QuestionHistory matches snapshot', () => {
    const { container } = render(<QuestionHistory changeFlow={changeFlow} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('changeQuestion should pop off all values untill the current question is no longer in the stack flow ', () => {
    const currentQuestion = ['DE20'];
    flowPush(['DE34']);
    flowPush(['DE20']);
    flowPush(['DE15']);
    changeQuestion(currentQuestion, changeFlow);
    expect(flowPeek()).toEqual(['DE20']);
  });
});
