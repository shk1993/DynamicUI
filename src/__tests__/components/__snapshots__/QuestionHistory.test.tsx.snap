// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionHistory component checks whether the QuestionHistory matches snapshot 1`] = `
<div>
  <p>
    <button
      class="MuiButtonBase-root-17 MuiButton-root-1 MuiButton-sizeSmall-14"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label-2"
      >
        DE30‚ü∂ 
      </span>
      <span
        class="MuiTouchRipple-root-20"
      />
    </button>
  </p>
</div>
`;
